Update GPU section of the documentation (#4816)
AMReX-Codes/amrex: Docs/sphinx_documentation/source/GPU.rst

(HIP for non-AMD systems and SYCL for non Intel systems).  It may work with
a given version, but AMReX does not track or guarantee such functionality.

AMReX uses an ``MPI+X`` approach to hierarchical parallelism. When running on
CPUs, ``X`` is ``OpenMP``, and threads are used to process tiles assigned to the
same MPI rank concurrently, as detailed in :ref:`sec:basics:mfiter:tiling`. On GPUs,
``X`` is one of ``CUDA/HIP/SYCL``, and tiling is disabled by default
to mitigate the overhead associated with kernel launching. Instead, kernels are usually
launched at the ``Box`` level, and one or more cells

> in a given ``Box`` are mapped to a each GPU thread, as detailed in :numref:`fig:gpu:threads`

Reviewer: `a each` needs to be fixed.

below.

Presented here is an overview of important features of AMReX's GPU strategy.
Additional information that is required for creating GPU applications is
detailed throughout the rest of this chapter:

- Each MPI rank offloads its work to a single GPU. Multiple ranks can share the
  same device, but for best performance we usually recommend ``(MPI ranks == Number of GPUs)``.

- To provide performance portability, GPU kernels are usually launched through ``ParallelFor`` looping constructs
  that use GPU extended lambdas to specify the work to be performed on each loop element. When compiled with GPU

>   support, these constructs launch kernels with a large number GPU threads that only work on a few cells each. This work

Reviewer: Missing `of` in `a large number GPU threads`?

   |   The lo and hi of one tiled box are marked.        |   thread, each thread using a box with lo = hi.      |
   +-----------------------------------------------------+------------------------------------------------------+

- These kernels are usually launched inside AMReX's :cpp:`MFIter` and :cpp:`ParIter`
  loops, since in AMReX's approach to parallelism it is assumed that separate ``Box`` objects

>   can be processed independently. However, AMReX also provides a ``MultiFab`` version

Reviewer: For consistency, should we add `:cpp` to names like `MultiFab`?

>   of ``ParallelFor`` that can process and entire level worth of ``Box`` objects in

Reviewer: ```suggestion
  of ``ParallelFor`` that can process an entire level worth of ``Box`` objects in
```

.. _sec:gpu:particle:

As with ``MultiFab``, particle data stored in AMReX ``ParticleContainer`` classes can be
stored in GPU-accessible memory when AMReX is compiled with ``USE_CUDA=TRUE``, ``USE_HIP=TRUE``,
or ``USE_SYCL=TRUE``.  The type of memory used by a given ``ParticleContainer`` can be controlled
by the ``Allocator`` template parameter. By default, when compiled with GPU support ``ParticleContainer`` uses ``The_Arena()``. This means that the :cpp:`dataPtr` associated with particle data
can be passed into GPU kernels. These kernels can be launched with a variety of approaches,
including AMReX's native kernel launching mechanisms as well OpenMP and OpenACC. Using AMReX's C++ syntax, a kernel launch involving particle data might look like:

.. highlight:: c++

::

   for(MyParIter pti(pc, lev); pti.isValid(); ++pti)
   {
       auto& ptile = pti.GetParticleTile();
       auto ptd = tile.getParticleTileData();
       const auto np = tile.numParticles();
       amrex::ParallelFor( np,
       [=] AMREX_GPU_DEVICE (const int ip) noexcept
       {

>            ptd[i].make_invalid();

Reviewer 2: ```suggestion
           ptd.id(i).make_invalid();
```

.. _sec:gpu:particle:

As with ``MultiFab``, particle data stored in AMReX ``ParticleContainer`` classes can be

> stored in GPU-accessible memory when AMReX is compiled with ``USE_CUDA=TRUE``, ``USE_HIP=TRUE``,

Reviewer: `USE_CUDA` etc. are for GNU Make only. Maybe we should avoid them here.

> or ``USE_SYCL=TRUE``.  The type of memory used by a given ``ParticleContainer`` can be controlled

Reviewer: ```suggestion
stored in GPU-accessible memory when AMReX is compiled with GPU support.  The type of memory used by a given ``ParticleContainer`` can be controlled
```
